apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ClientPool.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Grpc.Auth;
+
+namespace Google.Cloud.Spanner.V1
+{
+    internal static class ClientPool
+    {
+        static readonly ConcurrentDictionary<ITokenAccess, ClientEntry> ClientEntryPool =
+                new ConcurrentDictionary<ITokenAccess, ClientEntry>();
+
+        public static async Task<SpannerClient> AcquireClientAsync(ITokenAccess credentials)
+        {
+            ClientEntry entry = ClientEntryPool.GetOrAdd(credentials ?? ApplicationDefault.Instance, key => new ClientEntry(key));
+            return await entry.AcquireClientFromEntryAsync();
+        }
+
+        public static int Timeout { get; set; }
+
+        /// <summary>
+        /// Callers should be careful not to call AcquireClient while CloseAll is still running.
+        /// It is possible to obtain a stale client.
+        /// </summary>
+        /// <returns></returns>
+        public static async Task CloseAllAsync()
+        {
+            await SpannerClient.ShutdownDefaultChannelsAsync();
+            ClientEntryPool.Clear();
+        }
+
+        public class ClientEntry
```

Comment:
Presumably *all* the public members here could be internal - I think it would be good to make them explicitly internal, so readers don't need to check the access of the outer class.